### PR TITLE
Warn when a plugin does not proto3 optional instead of error-ing

### PR DESCRIPTION
**Pull request by doriable**
This changes the behaviour of `buf generate` to warn when
a feature is not supported instead of erroring. And only error
in cases when edition is found.
This change is to keep backwards compatibility with pre-1.32.0
versions of the CLI, which would warn but ignore unsupported
features and continue to generate code.

### CODE REVIEW

File: private/buf/bufgen/features.go [hunk 1]
```diff
@@ -132,11 +134,22 @@ func checkRequiredFeatures(
 			sort.Slice(failedFeatures, func(i, j int) bool {
 				return failedFeatures[i] < failedFeatures[j]
 			})
+			// We only log failed features (whereas we error on failed editions). This is in
+			// keeping with CLI versions pre-1.32.0 (BSR-3931).
+			_, _ = fmt.Fprintf(stderrContainer.Stderr(), "Warning: plugin %q does not support required features.\n", pluginName)
```
[the remark this answers is not in the document]
Comment (doriable) [reply private/buf/bufgen/features.go:139]: Oh, it was mostly because I was following the older pattern with the multi-line warning.
But I can adjust this to use `logger.Warn` and construct a single log-line.

File: private/buf/bufgen/features.go [hunk 2]
```diff
@@ -132,11 +134,22 @@ func checkRequiredFeatures(
 			sort.Slice(failedFeatures, func(i, j int) bool {
 				return failedFeatures[i] < failedFeatures[j]
 			})
+			// We only log failed features (whereas we error on failed editions). This is in
+			// keeping with CLI versions pre-1.32.0 (BSR-3931).
```
[the remark this answers is not in the document]
Comment (timostamm) [reply private/buf/bufgen/features.go:138]: Under the hood, proto3 optional fields are put into a synthetic `oneof`. As a result, the code generated by a plugin that does not support proto3 optional will still serialize correctly. 

Code generated by a plugin that does not support editions will most likely behave much worse. In practice, the plugin will likely print an error and exit with an error code (or populate `CodeGeneratorResponse.error`). But there is no guarantee that it does. It could just as well generate code that looks okay, but crashes at runtime, or serializes in a non-conformant way.

Even if we chose to keep `FEATURE_PROTO3_OPTIONAL` as a warning for BC, I think we need to error for `FEATURE_SUPPORTS_EDITIONS`.

[the remark this answers is not in the document]
Comment (doriable) [reply private/buf/bufgen/features.go:138]: > I would really prefer to see a config setting to treat the error as a warning for a particular plugin+feature pair. That way we could make it default to being an error for v2 config (and buf migrate could auto-add the opt-out flag when migrating buf.gen.yaml, to preserve back compat).

I don't think that's too controversial, tbh, but it might take a little bit to implement, not necessarily for complexity reasons, but we want to make sure we capture a reasonable UX for the configs without breaking what we currently have released.

Given the comments in this thread, I'm going to take the route or warning on only on `FEATURE_PROTO3_OPTIONAL` for now, since I do think we want to move things closer to a safer model.

File: private/buf/bufgen/features.go [hunk 3]
```diff
@@ -85,6 +86,7 @@ func computeRequiredFeatures(image bufimage.Image) *requiredFeatures {
 }
 
 func checkRequiredFeatures(
+	container appext.LoggerContainer,
```
Comment (bufdev) [inline private/buf/bufgen/features.go:89]: Just pass in a `*zap.Logger` that you got from `NewGenerator` and revert the change to `appext.Container`

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 ## [Unreleased]
 
-- No changes yet.
+- Update `buf generate` to warn when proto3 optional is required by a file but not supported
```
[the remark this answers is not in the document]
Comment (doriable) [reply CHANGELOG.md:5]: Yeah, I can remove for now -- I just wasn't sure if we wanted a changelog for this, so I put it up.

**Review by bufdev — APPROVED** [review]

**Review by pkwarren — APPROVED** [review]
